sshmux.go

```diff
@@ -401,6 +416,50 @@ func sshmuxServer(configFile string) {
 	}
 }
 
+func sshmuxServer(configFile string) {
+	configFileBytes, err := os.ReadFile(configFile)
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = json.Unmarshal(configFileBytes, &config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	sshConfig := &ssh.ServerConfig{
+		ServerVersion:           "SSH-2.0-taokystrong",
+		PublicKeyAuthAlgorithms: ssh.DefaultPubKeyAuthAlgos(),
+	}
+	for _, keyFile := range config.HostKeys {
+		bytes, err := os.ReadFile(keyFile)
+		if err != nil {
+			log.Fatal(err)
+		}
+		key, err := ssh.ParsePrivateKey(bytes)
+		if err != nil {
+			log.Fatal(err)
+		}
+		sshConfig.AddHostKey(key)
+	}
+	waitgroup := sync.WaitGroup{}
+	waitgroup.Add(1)
+	if config.Address == config.ProxiedAddress {
+		if config.Address == "" {
+			log.Println("No address specified, defaulting to 0.0.0.0:8022")
+			go sshmuxListenAddr("0.0.0.0:8022", &waitgroup, sshConfig, false, false)
+		} else {
+			go sshmuxListenAddr(config.Address, &waitgroup, sshConfig, true, true)
+		}
+	} else {
+		if config.Address != "" {
+			go sshmuxListenAddr(config.Address, &waitgroup, sshConfig, false, false)
+		}
+		if config.ProxiedAddress != "" {
+			go sshmuxListenAddr(config.ProxiedAddress, &waitgroup, sshConfig, true, false)
+		}
+	}
+	waitgroup.Wait()
```

Comment:
The waitgroup is only added by 1, but in the `else` clause of the outer `if`, it's possible that `wg.Done()` may be called twice. This will cause `wg.Done()` to panic.

Reply:
It’s by design that we want the program to quit on either of the listeners fail. After any of the `wg.Done()` call, `wg.Wait()` will be unblocked and `main` will quit.

If we want to resolve the problem, we may need to add another semaphore for synchronizing the state…

It adds so much complexity that I would prefer the IP allowlist/blocklist way so that we can reuse the same port (and thus listener) for `PROXY` and normal sessions.